src/java/com/unifina/signalpath/blockchain/EthereumJsonRpc.java

```diff
@@ -2,85 +2,33 @@
 
 import com.google.common.collect.ImmutableMap;
 import com.google.gson.Gson;
-import com.mashape.unirest.http.HttpResponse;
-import com.mashape.unirest.http.JsonNode;
 import com.mashape.unirest.http.Unirest;
-import com.mashape.unirest.http.exceptions.UnirestException;
 import com.mashape.unirest.request.body.RequestBodyEntity;
 import org.json.JSONObject;
 
 import java.util.List;
 
-class EthereumJsonRpc {
-	private static final int CALL_ID = 123;
-	private final String url;
-
-	EthereumJsonRpc(String url) {
+public abstract class EthereumJsonRpc {
+	protected final String url;
+	protected JsonRpcResponseHandler handler;
+	EthereumJsonRpc(String url, JsonRpcResponseHandler handler) {
 		this.url = url;
+		this.handler = handler;
 	}
 
-	JSONObject rpcCall(String method, List params) throws UnirestError, HttpStatusError, ErrorObjectError {
-		try {
-			HttpResponse<JsonNode> response = formRequest(method, params, CALL_ID).asJson();
-
-			if (statusCodeIsNot2XX(response.getStatus())) {
-				throw new HttpStatusError(response.getStatus(), response.getBody());
-			}
-
-			JSONObject responseJson = response.getBody().getObject();
-			if (responseJson.opt("error") != null) {
-				throw new ErrorObjectError(responseJson.optJSONObject("error"));
-			}
-
-			return responseJson;
-		} catch (UnirestException e) {
-			throw new UnirestError(e);
-		}
-	}
-
-	private RequestBodyEntity formRequest(String method, List params, int callId) {
-		return Unirest.post(url)
-				.header("Content-Type", "application/json")
-				.body(new Gson().toJson(ImmutableMap.of(
-			"id", callId,
-			"jsonrpc", "2.0",
-			"method", method,
-			"params", params
-		)));
-	}
+	/*
+		implementation should call handler.processResponse()
+	 */
+	public abstract void rpcCall(String method, List params, int callId) throws Exception;
```

Comment:
More specific exception type?

Reply:
well the rpc could, in theory, go over a number of channels: http, websocket, unix socket file, etc. Specific implementations throw different exceptions (eg HttpException). So I would keep the Exception generic.